### ISSUE TITLE
Bug? `NoReturn`, `overload`, `Literal[False]`

### ISSUE THREAD

**Issue by smackesey** [microsoft/pyright#2700]
This was originally posted to [StackOverflow](https://stackoverflow.com/questions/70327505/python-type-checking-literalfalse-overload-and-noreturn), but reposting here since a commenter said `mypy` does not have the same issue. Not 100% sure it's a bug, but here goes (using pyright 1.1.195):

I have the following (typed) Python function:

```python
from typing import overload, Literal, NoReturn

@overload
def check(condition: Literal[False], msg: str) -> NoReturn:
    pass

@overload
def check(condition: Literal[True], msg: str) -> None:
    pass

def check(condition, msg):
    if not condition:
        raise Exception(msg)
```

Pyright complains:

```
Overloaded function implementation is not consistent with signature of overload 1
  Function return type "NoReturn" is incompatible with type "None"
    Type cannot be assigned to type "None"
```

Is this a bug? I would think Pyright should be able to determine that `check` will always throw an error if `condition` is `False`.

**Comment by erictraut** [comment]
No, this is intended behavior. It falls out of pyright's rules for return type inference. Pyright will infer a `NoReturn` type only in the case that all code paths raise an exception. It's quite common for some code paths raise an exception, and it would produce many false positives if pyright were to include `NoReturn` in the union in that case, so `NoReturn` is always elided from a union in an inferred return type.

Mypy doesn't have any support for return type inference, so that explains why this doesn't occur for mypy.

The correct way to annotate this is to include an explicit `None | NoReturn` return type for the implementation.

**Comment by smackesey** [comment]
Doesn't including `None | NoReturn` for the implementation defeat the purpose of using an `overload` here in the first place? It appears that Pyright is not matching my overload signature when I do this:

```python
from typing import overload, Literal, NoReturn, Union

@overload
def check(condition: Literal[False], msg: str) -> NoReturn:
    pass

@overload
def check(condition: Literal[True], msg: str) -> None:
    pass

def check(condition, msg) -> Union[None, NoReturn]:
    if not condition:
        raise Exception(msg)

# Type Error: Function with declared type of "int" must return value
def test_check(x: int) -> int:
    if x > 5:
        return x
    else:
        # Pyright doesn't know this will never return.
        check(False, 'This is an error.')
```

**Comment by erictraut** [comment]
We're talking about two different issues here. 

Issue 1: What type annotation do you need to provide for the implementation so pyright doesn't complain about the implementation being incompatible with the overloads.

Issue 2: When you call the function, does pyright use the `NoReturn` overload result to determine that the code path never falls through, resulting in an implied `None` return.

Your initial bug report was about issue 1. This is not a bug, it is intended behavior.

As for issue 2, a type checker should be able to determine in this case that a fall-through doesn't occur. This is completely separate from the initial issue above. I'd normally ask you to file a separate bug for this issue, but I'll save you the time. Unfortunately, pyright cannot make this determination because of its core architecture. The "reachability" of nodes within a code flow graph cannot rely on type evaluations because type evaluations depend on reachability of code flow nodes. To work around this chicken-and-egg problem, the logic that determines reachability does some basic checks to determine if the called function is potentially a "NoReturn" function, but these basic checks are not sophisticated enough to handle overload evaluation. Evaluating overloads requires the full type evaluator, and that depends on reachability.

To work around this, you can add an `assert False` or adopt a different approach for your `check`. It's a bit odd that you'd have a `check` function that accepts only literal `True` of `False`. If you know whether `condition` is true or false statically, then what is the use of a function that dynamically raises an exception? Why not just raise the exception directly in the case where the condition is false?

**Comment by smackesey** [comment]
Thanks-- the `check` definition I presented is a simplification of the actual code I'm working with. For my Stack Overflow question I was trying to discern the underlying principle.

The actual "check" function (which I'm not at liberty to change, aside from adding type annotations) is this:

```python
def _cli_load_invariant(condition, msg=None):
    msg = (
        msg
        or "Invalid set of CLI arguments for loading repository/pipeline. See --help for details."
    )
    if not condition:
        raise UsageError(msg)
```

My aim was to add annotations to this so that Pyright would understand when it would throw an error. From your explanation, sounds like that's not possible though, so I guess I'll stick an `assert False` in the necessary location.
